### ISSUE TITLE
Unable to update a state in parent widget

### ISSUE THREAD

**Issue by edgebasis** [felangel/bloc#1128]
How can i update state in the parent widget? It fetches from API at first and then when i update the API from another screen it does not reflect on the parent screen. It refreshes or updates only if i do a hot reload or save the file to reload. But when i update the api, it does not reflect. 
I have gone through some of the discussions here and it appears that I might be using the equitable wrong. I tried to play with that but did not reach anywhere.
I provided the bloc instance only once at the parent and then used the same bloc to the child screen. I am going to post the model, event and states files since the UI is too long..

model:

```
class Order extends Equatable{
  int id;
  User user;
  String refCode;
  String createdAt;
  String updatedAt;
  bool ordered;
  String status;
  String orderType;
  Address address;
  String payment;
  String coupon;
  bool beingDelivered;
  bool received;
  bool refundRequested;
  bool refundGranted;
  double total;
  int totalItemQuantity;
  List<OrderItems> orderItems;
  String error;

  Order(
      {this.id,
      this.user,
      this.refCode,
      this.createdAt,
      this.updatedAt,
      this.ordered,
      this.status,
      this.orderType,
      this.address,
      this.payment,
      this.coupon,
      this.beingDelivered,
      this.received,
      this.refundRequested,
      this.refundGranted,
      this.total,
      this.totalItemQuantity,
      this.orderItems});

  Order.fromJson(Map<String, dynamic> json) {
    id = json['id'];
    user = json['user'] != null ? new User.fromJson(json['user']) : null;
    refCode = json['ref_code'];
    createdAt = json['created_at'];
    updatedAt = json['updated_at'];
    ordered = json['ordered'];
    status = json['status'];
    orderType = json['order_type'];
    address =
        json['address'] != null ? new Address.fromJson(json['address']) : null;
    payment = json['payment'];
    coupon = json['coupon'];
    beingDelivered = json['being_delivered'];
    received = json['received'];
    refundRequested = json['refund_requested'];
    refundGranted = json['refund_granted'];
    total = json['total'];
    totalItemQuantity = json['total_item_quantity'];
    if (json['order_items'] != null) {
      orderItems = new List<OrderItems>();
      json['order_items'].forEach((v) {
        orderItems.add(new OrderItems.fromJson(v));
      });
    }
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = new Map<String, dynamic>();
    data['id'] = this.id;
    if (this.user != null) {
      data['user'] = this.user.toJson();
    }
    data['ref_code'] = this.refCode;
    data['created_at'] = this.createdAt;
    data['updated_at'] = this.updatedAt;
    data['ordered'] = this.ordered;
    data['status'] = this.status;
    data['order_type'] = this.orderType;
    if (this.address != null) {
      data['address'] = this.address.toJson();
    }
    data['payment'] = this.payment;
    data['coupon'] = this.coupon;
    data['being_delivered'] = this.beingDelivered;
    data['received'] = this.received;
    data['refund_requested'] = this.refundRequested;
    data['refund_granted'] = this.refundGranted;
    data['total'] = this.total;
    data['total_item_quantity'] = this.totalItemQuantity;
    if (this.orderItems != null) {
      data['order_items'] = this.orderItems.map((v) => v.toJson()).toList();
    }
    return data;
  }

  Order.withError(String errorValue) : error = errorValue;

  @override
  // TODO: implement props
  List<Object> get props => [this.id,
    this.user,
    this.refCode,
    this.createdAt,
    this.updatedAt,
    this.ordered,
    this.status,
    this.orderType,
    this.address,
    this.payment,
    this.coupon,
    this.beingDelivered,
    this.received,
    this.refundRequested,
    this.refundGranted,
    this.total,
    this.totalItemQuantity,
    this.orderItems];
}
```

bloc:

```

class CartBloc extends Bloc<CartEvent, CartState>{
  MenuItemsRepository menuItemsRepository;


  CartBloc({@required this.menuItemsRepository});

  @override
  // TODO: implement initialState
  CartState get initialState => CartInitial();

  @override
  Stream<CartState> mapEventToState(CartEvent event) async* {
    if(event is FetchOrderSummaryEvent){
      yield CartLoading();
      try {
        Order order = await menuItemsRepository.getOrder();
        yield CartLoaded(order: order);
      } catch(err){
        yield CartFailure(message: err.message);
      }
    }
    if(event is AddItemToCartEvent){
      yield CartLoading();
      try{
        await menuItemsRepository.addItemToCart(event.slug, event.variations);
        try {
          Order order = await menuItemsRepository.getOrder();


          yield CartLoaded(order: order);

        }catch(e){
          yield CartFailure(message: e.toString());
        }
      } catch(err){
        yield CartFailure(message: err.toString());
      }
    }
    if(event is RemoveItemFromCartEvent){
      yield CartLoading();
      try{
        await menuItemsRepository.removeItemFromCart(event.slug);
        try {
          Order order = await menuItemsRepository.getOrder();
          yield CartLoaded(order: order);
        }catch(e){
          yield CartFailure(message: e.toString());
        }
      } catch(err){
        yield CartFailure(message: err.toString());

      }
    }
  }

}
```

state:

```
@immutable
abstract class CartState extends Equatable {
  CartState();

  @override
  List<Object> get props => [];
}

class CartInitial extends CartState {
  @override
  // TODO: implement props
  List<Object> get props => [];
}

class CartLoading extends CartState {
  @override
  // TODO: implement props
  List<Object> get props => [];
}

class CartLoaded extends CartState {
  Order order;
  CartLoaded({@required this.order});


  @override
  // TODO: implement props
  List<Object> get props => [order];
}

class CartFailure extends CartState {
  String message;

  CartFailure({@required this.message});

  @override
  // TODO: implement props
  List<Object> get props => [message];
}
```

event:

```
@immutable
abstract class CartEvent extends Equatable {
  CartEvent();

@override
List<Object> get props => [];
}

class FetchOrderSummaryEvent extends CartEvent{

}
class RemoveItemFromCartEvent extends CartEvent {
  String slug;


  RemoveItemFromCartEvent({@required this.slug});

  @override
  List<Object> get props => [];
}

class AddItemToCartEvent extends CartEvent {
  String slug;
  List<int> variations;

  AddItemToCartEvent({@required this.slug, @required this.variations});

  @override
  List<Object> get props => [];
}

class RemoveItemEvent extends CartEvent {
  @override
  List<Object> get props => [];
}

class CheckoutEvent extends CartEvent {
  @override
  List<Object> get props => [];
}

```

Please let me know how this can be solved.
thanks

**Comment by felangel** [comment]
Hi @edgebasis 👋 
Thanks for opening an issue!

Are you able to share a link to a sample app which illustrates the problem you're having? It's much easier for me to help if I can run/debug the application locally. Thanks 👍

**Comment by edgebasis** [comment]
thanks @felangel ,
here is the repo
https://github.com/edgebasis/flutterApp

**Comment by heshaShawky** [comment]
This is related to my issue here #1091 the problem is that you are creating a new object every time you trigger an event. 

you have to use the state data the first time and add the new item from your API after its success to the list. 

Check this code: 
https://github.com/heshaShawky/Adding-new-item-to-the-list-issue-repository/blob/master/lib/comments_bloc/comments_bloc.dart

**Comment by edgebasis** [comment]
Yes but I am calling an Api each time there is an update in the shopping
cart. So if I add an item I need to call the api as I am not using any
local storage nor state for the cart. So update api then fetch the list of
items in the cart from api..

**Comment by edgebasis** [comment]
I tried to trace the state in the parent widget and i could see that the data the state holds is always old. Since I fetch the data from  the api in the child widget and parent widget

**Comment by edgebasis** [comment]
any solution for this?

**Comment by RollyPeres** [comment]
Hi @edgebasis 👋

Tried to run your sample locally but I'm getting a quite a few dio and http errors due to invalid signatures. You mind migrating your sample to latest flutter stable version? Might be a good idea to even try and recreate your issue in a smaller sample which eliminates all the unneeded dependencies and focuses on flutter bloc. 👍

**Comment by edgebasis** [comment]
Hi @RollyPeres , i could find out the problem "finally!" and for me it was easier to dig more than change the code base and make another version to focus on the bloc or the issue..

I am gonna put down the problem cause and the solution and hopefully helps others who might run into this issue like i did:

okay, my problem was. that i was re-instantiating the bloc repository in a child screen:

```
class OrderSummary extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return BlocProvider(
      create: (context) => CartBloc(menuItemsRepository: MenuItemsRepository()),
      child: ShowOrderSummary(),
    );
  }
}
```
this was making the calls to the API normal but was making a separate instance of the bloc in the child rather than use the same bloc instance inherited from the parent. I am sure i have read this somewhere mentioned as a solution for another issue in here where @felangel was showing a solution to another problem but as always its a simple 1 line of code that can make you sleepless for days or weeks!
so, the problem was that there was a new instance of the bloc being provided in the child rather than using the same parent one.. as soon as i removed the above lines of code in the child, the cart in the parent was updating smoothly..

thanks to everyone who tried to help and thank you @felangel for providing this platform...

**Comment by RollyPeres** [comment]
@edgebasis glad you figured it out and thanks for sharing your findings, might indeed be helpful to others 👍
